File: src/org/jgll/util/SPPFToDot.java
package org.jgll.util;

import static org.jgll.util.GraphVizUtil.EDGE;
import static org.jgll.util.GraphVizUtil.INTERMEDIATE_NODE;
import static org.jgll.util.GraphVizUtil.PACKED_NODE;
import static org.jgll.util.GraphVizUtil.SYMBOL_NODE;

import org.jgll.sppf.IntermediateNode;
import org.jgll.sppf.ListSymbolNode;
import org.jgll.sppf.NonterminalSymbolNode;
import org.jgll.sppf.PackedNode;
import org.jgll.sppf.SPPFNode;
import org.jgll.sppf.TerminalSymbolNode;
import org.jgll.traversal.SPPFVisitorUtil;
import org.jgll.traversal.SPPFVisitor;

/**
 * Creates a Graphviz's dot format representation of an SPPF node.
 * 
 * @author Ali Afroozeh
 * 
 * @see SPPFVisitor
 */
public class SPPFToDot extends ToDot implements SPPFVisitor  {
	
	private final boolean showPackedNodeLabel;
	protected StringBuilder sb;
	
	public SPPFToDot() {
		this(false);
	}
	
	public SPPFToDot(boolean showPackedNodeLabel) {
		this.showPackedNodeLabel = showPackedNodeLabel;
		this.sb = new StringBuilder();
	}

	@Override
	public void visit(TerminalSymbolNode node) {
		if(!node.isVisited()) {
			node.setVisited(true);
			String label = node.getLabel();
			// Replaces the Java-style unicode char for epsilon with the graphviz one
			label.replace("\u03B5", "&epsilon;");
			sb.append("\"" + getId(node) + "\"" + String.format(SYMBOL_NODE, replaceWhiteSpace(label)) + "\n");
		}
	}

	@Override
	public void visit(NonterminalSymbolNode node) {
		if(!node.isVisited()) {
			node.setVisited(true);
	
			sb.append("\"" + getId(node) + "\"" + String.format(SYMBOL_NODE, replaceWhiteSpace(node.getLabel())) + "\n");
			addEdgesToChildren(node);
			
			SPPFVisitorUtil.visitChildren(node, this);
		}
	}

	@Override
	public void visit(IntermediateNode node) {
		if(!node.isVisited()) {
			node.setVisited(true);
	
			sb.append("\"" + getId(node) + "\"" + String.format(INTERMEDIATE_NODE, replaceWhiteSpace(node.toString())) + "\n");
			addEdgesToChildren(node);
	
			SPPFVisitorUtil.visitChildren(node, this);
		}
	}

	@Override
	public void visit(PackedNode node) {
		if(!node.isVisited()) {
			node.setVisited(true);
	
			if(showPackedNodeLabel) {
				sb.append("\"" + getId(node) + "\"" + String.format(PACKED_NODE, replaceWhiteSpace(node.toString())) + "\n");
			} else {
				sb.append("\"" + getId(node) + "\"" + String.format(PACKED_NODE, "") + "\n");
			}
			addEdgesToChildren(node);
			
			SPPFVisitorUtil.visitChildren(node, this);
		}
	}
	
	protected void addEdgesToChildren(SPPFNode node) {
		for (SPPFNode child : node.getChildren()) {
			addEdgeToChild(node, child);
		}
	}
	
	protected void addEdgeToChild(SPPFNode parentNode, SPPFNode childNode) {
		sb.append(EDGE + "\"" + getId(parentNode) + "\"" + "->" + "{\"" + getId(childNode) + "\"}" + "\n");
	}
	
	protected String replaceWhiteSpace(String s) {
		return s.replace("\\", "\\\\").replace("\t", "\\\\t").replace("\n", "\\\\n").replace("\r", "\\\\r").replace("\"", "\\\"");
	}

	@Override
	public void visit(ListSymbolNode node) {
		visit((NonterminalSymbolNode)node);
	}
	
	public String getString() {
		return sb.toString();
	}
}

File: src-scala/org/jgll_staged/util/dot/SPPFToDot.scala
package org.jgll_staged.util.dot

import org.jgll_staged.util.dot.GraphVizUtil.EDGE
import org.jgll_staged.util.dot.GraphVizUtil.INTERMEDIATE_NODE
import org.jgll_staged.util.dot.GraphVizUtil.PACKED_NODE
import org.jgll_staged.util.dot.GraphVizUtil.SYMBOL_NODE
import org.jgll_staged.sppf.IntermediateNode
import org.jgll_staged.sppf.ListSymbolNode
import org.jgll_staged.sppf.NonterminalSymbolNode
import org.jgll_staged.sppf.PackedNode
import org.jgll_staged.sppf.SPPFNode
import org.jgll_staged.sppf.TerminalSymbolNode
import org.jgll_staged.traversal.SPPFVisitorUtil
import org.jgll_staged.traversal.SPPFVisitor
//remove if not needed
import scala.collection.JavaConversions._

class SPPFToDot(private val showPackedNodeLabel: Boolean) extends ToDot with SPPFVisitor {

  protected var sb: StringBuilder = new StringBuilder()

  def this() {
    this(false)
  }

  override def visit(node: TerminalSymbolNode) {
    if (!node.isVisited) {
      node.setVisited(true)
      val label = node.getLabel
      label.replace("ε", "&epsilon;")
      sb.append("\"" + getId(node) + "\"" + 
        String.format(SYMBOL_NODE, replaceWhiteSpace(label)) + 
        "\n")
    }
  }

  override def visit(node: NonterminalSymbolNode) {
    if (!node.isVisited) {
      node.setVisited(true)
      sb.append("\"" + getId(node) + "\"" + 
        String.format(SYMBOL_NODE, replaceWhiteSpace(node.getLabel)) + 
        "\n")
      addEdgesToChildren(node)
      SPPFVisitorUtil.visitChildren(node, this)
    }
  }

  override def visit(node: IntermediateNode) {
    if (!node.isVisited) {
      node.setVisited(true)
      sb.append("\"" + getId(node) + "\"" + 
        String.format(INTERMEDIATE_NODE, replaceWhiteSpace(node.toString)) + 
        "\n")
      addEdgesToChildren(node)
      SPPFVisitorUtil.visitChildren(node, this)
    }
  }

  override def visit(node: PackedNode) {
    if (!node.isVisited) {
      node.setVisited(true)
      if (showPackedNodeLabel) {
        sb.append("\"" + getId(node) + "\"" + 
          String.format(PACKED_NODE, replaceWhiteSpace(node.toString)) + 
          "\n")
      } else {
        sb.append("\"" + getId(node) + "\"" + String.format(PACKED_NODE, "") + 
          "\n")
      }
      addEdgesToChildren(node)
      SPPFVisitorUtil.visitChildren(node, this)
    }
  }

  protected def addEdgesToChildren(node: SPPFNode) {
    for (child <- node.getChildren) {
      addEdgeToChild(node, child)
    }
  }

  protected def addEdgeToChild(parentNode: SPPFNode, childNode: SPPFNode) {
    sb.append(EDGE + "\"" + getId(parentNode) + "\"" + "->" + "{\"" + 
      getId(childNode) + 
      "\"}" + 
      "\n")
  }

  protected def replaceWhiteSpace(s: String): String = {
    s.replace("\\", "\\\\").replace("\t", "\\\\t").replace("\n", "\\\\n")
      .replace("\r", "\\\\r")
      .replace("\"", "\\\"")
  }

  override def visit(node: ListSymbolNode) {
    visit(node.asInstanceOf[NonterminalSymbolNode])
  }

  def getString(): String = sb.toString
}
